tests/graph/test_basic.py

```diff
@@ -749,3 +749,22 @@ def test_truncated_graph_inputs():
     assert truncated_graph_inputs([o2], [y2, z]) == [y2]
     # Disconnected output is present
     assert truncated_graph_inputs([o2, z], [y2]) == [z, y2]
+
+
+def test_truncated_graph_inputs_repeated_input():
+    """Test that truncated_graph_inputs does not return repeated inputs."""
+    x = MyVariable(1)
+    x.name = "x"
+    y = MyVariable(1)
+    y.name = "y"
+
+    trunc_inp1 = MyOp(x, y)
+    trunc_inp1.name = "trunc_inp1"
+
+    trunc_inp2 = MyOp(x, y)
+    trunc_inp2.name = "trunc_inp2"
+
+    o = MyOp(trunc_inp1, trunc_inp1, trunc_inp2, trunc_inp2)
```

Comment:
Order is graph traversal dependent. Closer nodes depth first appear first

Reply:
The correct should not include `y`

Reply:
A subgraph must have independent inputs to be valid. `trunc_inp` depends on `x`, so they can't both be inputs?

Reply:
`trunc_inp` is declared as `ancestors_to_include`, therefore, other added variables should not depend on them. Yes,  `trunc_inp` depends on `x` but not vice versa and `x` is absolutely valid to include. All the above dependencies are assumed to be totally ignored.

Reply:
Hmm, wouldn't it be more useful to return the true minimum inputs? In this case `[x, y]`. That's what I think I would want this function to do.

Reply:
I think it will serve a different purpose, it is not the number of input nodes that is minimized, but the nodes in the graph. replacing `trunc_inp` with `y` will add you one more node to the graph that is contained within `truncated_inputs`

Reply:
Okay, I'll try to let that sink. Any what to do about the order issues? `y` should never appear right?

Reply:
> Okay, I'll try to let that sink. Any what to do about the order issues? `y` should never appear right?

The only requirement for the order: "Does not depend on `ancestors_to_include`, but on `set(ancestors_to_include)`". `y` should not appear

Reply:
This means we need to check candidates in topological order no?

Reply:
hmmm

Reply:
candidates start with outputs

Reply:
What affected in your test case is the order `op.inputs`, which is not topological, right?